pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -1661,3 +1726,50 @@ func (ctrl *PersistentVolumeController) findDeletablePlugin(volume *v1.Persisten
 	}
 	return plugin, nil
 }
+
+// obtain provisioner/deleter name for a volume
+func (ctrl *PersistentVolumeController) getProvisionerNameFromVolume(volume *v1.PersistentVolume) string {
+	plugin, err := ctrl.findDeletablePlugin(volume)
+	if err != nil {
+		return "N/A"
+	}
+	if plugin == nil || plugin.IsMigratedToCSI() {
+		// Either an external provisioner was used for provisioning
+		// or a plugin has been migrated to CSI.
+		// If an external provisioner was used, i.e., plugin == nil, instead of using
+		// the AnnDynamicallyProvisioned annotation value, use the storageClass's Provisioner
+		// field to avoid explosion of the metric in the cases like local storage provisioner
+		// tagging a volume with arbitrary provisioner names
+		storageClass := v1helper.GetPersistentVolumeClass(volume)
+		class, err := ctrl.classLister.Get(storageClass)
+		if err != nil {
+			return "N/A"
+		}
+		provisionerName := class.Provisioner
+		if plugin != nil {
+			// use the CSI name instead of in-tree name
+			provisionerName, err = ctrl.getCSINameFromIntreeName(class.Provisioner)
+			if err != nil {
+				return "N/A"
+			}
+		}
+		return provisionerName
+	}
+	return plugin.GetPluginName()
+}
```

Comment:
Can this function be simplified to:

```golang
func (ctrl *PersistentVolumeController) getProvisionerNameFromVolume(volume *v1.PersistentVolume) string {
	plugin, _ := ctrl.findDeletablePlugin(volume)
	if plugin == nil {
		storageClass := v1helper.GetPersistentVolumeClass(volume)
		class, err := ctrl.classLister.Get(storageClass)
		if err != nil {
			return "N/A"
		}
		return class.Provisioner
	}
	if plugin.IsMigratedToCSI() {
		// use the CSI name instead of in-tree name
		provisionerName, err = ctrl.getCSINameFromIntreeName(class.Provisioner)
		if err != nil {
			return "N/A"
		}
		return provisionerName
	}
	return plugin.GetPluginName()
}
```

Reply:
This is a minor nit though.. I don't mind either version of the function.

Reply:
updated to return provisioner when getCSINameFromIntreeName fails